---
title: "Lab 10: Population genomics"
author: Grace Lawson
date: 11.2.2020
output: html_document
---

# Part 1: Reading VCF Data

```{r}
library(vcfR)
```

```{r}
data(vcfR_example)
# Object vcf contains example VCF data - calling object name with no function invokes "show" method that prints summary info
vcf
```

## The meta region
The meta region contains info about the file, its creation, abbreviations. Each line begins with '##'. 
```{r}
strwrap(vcf@meta[1:7])
```

meta region may include long lines - queryMETA() used to summarize 
```{r}
queryMETA(vcf)
```
```{r}
queryMETA(vcf, element = 'DP')
```

## The fix region
fix region contains info for each variant 
* Columns 1 & 2 = CHROM & POS specify chromosome number and position on chromosome
* REF & ALT columns = reference & alternate allelic states 
  - when multiple allelic states are present, commas used
* QUAL column = summarize quality of each variant over all samples
* FILTER column = info on whether a variant has passed quality assessment

```{r}
head(getFIX(vcf))
```

## The gt region
gt (genotype) region contains info about each variant for each sample

```{r}
vcf@gt[1:6, 1:4]
```

## vcfR
Can read VCF format files into memory using function read.vcfR()

```{r}
library(pinfsc50)
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz")
```
```{r}
head(vcf)
```
```{r}
# Save as VCF file
write.vcf(vcf, "myVCFdata_filtered.vcf.gz")
```

## Exercises
1. How to find out more info about read.vcfR()
  * ?read.vcfR()
2. How to figure out what "AD" stands for?
```{r}
queryMETA(vcf, element = 'AD')
```

3. How to view the last few lines of fix data?
```{r}
tail(vcf@fix)
```

4. 
```{r}
plot(vcf)
```

5. How to query the sample names?
```{r}
colnames(vcf@gt)
```

# Part 2: Analysis of Genome Data
## Opening & examining dataset
```{r}
library('vcfR')
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz")
```
```{r}
vcf
```

## Converting VCF data to genlight object
R package adegenet is used for population genetic analysis - works on data structures called 'genlight' objects.
  * use function vcfR2genlight() to convert vcfR objects to genlight objects
```{r}
library(deldir)
library(adegenet)
```

```{r}
x <- vcfR2genlight(vcf)
```

```{r}
x
```
Genlight objects can only support variants with two alleles. Genotypes are also stored differently 

```{r}
# vcfR
gt <- extract.gt(vcf, element = "GT")
gt[c(2,6,18), 1:3]
```
```{r}
# genlight
t(as.matrix(x))[c(1,5,17), 1:3]
```

In VCF data the samples are in columns & variants in rows. In genlight objects samples are in rows & variants in columns. Can use transpose function (t()) to convert between states. 

In VCF data there is no population concept - use adegenet for population data.
```{r}
pop(x) <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
popNames(x)
```

Population designation consists of a vector of same length as # samples. 
- each element indicates which population each sample belongs to
- (as.factor()) function transforms vector into factor -> factor understands all the  elements names the same are part of the same group

Ploidy is treated differently between two objects
- in VCF data each variant is independent -> can be of different ploidy levels
- in genlight different samples may be of different ploidy levels but within each sample all loci must be of same ploidy level
```{r}
ploidy(x) <- 2
```

## Distance matrices
Can create a distance matrix from a genlight object:
```{r}
x.dist <- dist(x)
```

```{r}
x.dist <- poppr::bitwise.dist(x)
```

## chromR objects
### Using chromR to locate unusual features in a genome
vcfR includes a genome browser that can be used to combine different data types.

### Creating chromR objects
```{r}
library(pinfsc50)
```

```{r}
# Find files
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

# Input the files.
vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")

# Create a chromR object.
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)
```

```{r}
chrom
```

```{r}
# histograms of data summaries
plot(chrom)
```

```{r, fig.width= 10}
# display plot of data distributed along chromosomal coordinates
chromoqc(chrom, dp.alpha = 66)
```

### Processing chromR objects
```{r}
chrom <- proc.chromR(chrom, verbose = TRUE)
```

```{r}
plot(chrom)
```

```{r, fig.width=10}
chromoqc(chrom, dp.alpha = 66)
```

```{r, fig.width=10}
# Use data after quality control measures
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz", verbose = FALSE)
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)
chrom <- proc.chromR(chrom, verbose = FALSE)
chromoqc(chrom, dp.alpha = 66)
```

### Tabular summaries

```{r}
head(chrom@var.info)
```
```{r}
head(chrom@win.info)
```

## Genetic differentiation
Geneticists often report heterozygosity as a measure of genetic diversity - probability that two alleles randomly chosen from a population will be different (Simpson's Index).  
In vcfR the function genetic_diff() is used to measure population diversity & differentiation. 

```{r}
data(vcfR_example)
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))
```
- function returns chromosome & position of each variant 

```{r}
knitr::kable(head(myDiff[,16:19]))
```
```{r}
knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))
```

- can also summarize data with violin plots
```{r}
library(reshape2)
library(ggplot2)

dpf <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)
```
```{r}
p <- ggplot(dpf, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```

## Exercises Part II
1. PLot G'st by genomic position
```{r}
plot(getPOS(vcf), myDiff$Gprimest,  pch = 20, col = "#1E90FF44", xlab = "", ylab = "", ylim = c(0, 1), xaxt = "n")
axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))
```

2. What is wrong with this analysis?
There is a small sample size

3. How do you zoom in on a chromosome region?
```{r, fig.width=10}
chromoqc(chrom, dp.alpha = 66, xlim = c(2e05, 4e05))
```

4. Use queryMETA() to look for other data in the file that may be of interest.
```{r}
queryMETA(vcf)
```

